rerun_cpp/src/rerun/component_type_registry.cpp

```diff
@@ -0,0 +1,27 @@
+#include "component_type_registry.hpp"
+
+namespace rerun {
+
+    /// Returns the handle to the registered component type for the given descriptor.
+    ///
+    /// Registers the component type when first encountered.
+    Result<ComponentTypeHandle> ComponentTypeRegistry::get_or_register(
+        const ComponentDescriptor& descriptor, std::shared_ptr<arrow::DataType> arrow_datatype
+    ) {
+        std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
looking at this again I think we should actually use a [`shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex.html): clearly folks are running this method from multiple threads and almost always we have everything we need already registered.

Makes the logic ofc a little bit more complicated but I think it's alright, let me know what you think :)
Essentially first acquire a [`std::shared_lock`](https://en.cppreference.com/w/cpp/thread/shared_lock.html) then drop it again if no value was found, then attempt again with a [`std::unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock.html) and _potentially_ adding the element.